商品購入機能の実装

tomo1107 opened the pull request:
# What
## 実装内容
- 商品購入機能の実装
- 売却済みの商品は、画像上に『sold out』の文字が表示されるようになっていること（商品一覧機能、商品詳細機能）
- ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されないこと（商品詳細機能）
- ログイン状態の出品者以外のユーザーでも、売却済みの商品に対しては「購入画面に進む」ボタンが表示されないこと（商品詳細機能）
- ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移すること（商品情報編集機能）

## 動作確認結果(Gyazo画像)
- 正しく情報を記入すると、商品の購入ができる動画
https://gyazo.com/f26d414534f2de575de0917c7a391635
- 入力に問題のある状態では商品の購入はできず、エラーメッセージが出力される動画
https://gyazo.com/f8f83f31acd21bea94ef7b794d56d5f2
- ログアウト状態のユーザーが、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/26da816f455025413eb0a54044427839
- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
https://gyazo.com/8f70e7caef5fe1a3bf4351d23e246f13
- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/99b07c100305476f1ecc292ad547d2c1
- 売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品一覧機能実装時に未実装であった場合）https://gyazo.com/b5ad8b29256acdd0849b72767a6b2a38
- 売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/db2f8ef6b48992cee9e23fc039ae6e2b
- ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/50aeea3dacbeb8a351b2ee0d02f0072d
- ログイン状態の出品者以外のユーザーでも、売却済みの商品に対しては「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/1dc94583d3105a585747b0fc32f54811
- ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/9ada341841dd74cd13248a6855cb7aab
- テスト結果の画像
https://gyazo.com/5055e37df5c4d10eb204ff2eefe5ec3e

# Why
- 商品購入機能が正常に実装出来ているかをご確認お願いいたします。

tomo1107 committed:
商品購入機能の実装

shodai-ito commented on app/views/items/show.html.erb:
```diff
@@ -23,14 +23,14 @@
         <%= link_to "削除", item_path(@item.id), method: :delete, class:"item-destroy" %>
       <% else %>
         <%# 商品が売れていない場合はこちらを表示しましょう %>
-        <%= link_to "購入画面に進む", "#" ,class:"item-red-btn"%>
+        <%= link_to "購入画面に進む", item_purchases_path(@item.id) ,class:"item-red-btn"%>
         <%# //商品が売れていない場合はこちらを表示しましょう %>
```
実装が完了したコメントアウトは削除しておきましょう！

shodai-ito commented on spec/models/purchase_delivery_information_spec.rb:
```diff
@@ -0,0 +1,81 @@
+        @purchase_delivery_information.valid?
+        expect(@purchase_delivery_information.errors.full_messages).to include("Address can't be blank")
+      end
+      it 'buildingが空の場合でも、保存できること' do
+        @purchase_delivery_information.building = ''
+        expect(@purchase_delivery_information).to be_valid
+      end
```
こちらのテストは正常系のcontextの方に移動させましょう！

shodai-ito commented on spec/models/purchase_delivery_information_spec.rb:
```diff
@@ -0,0 +1,81 @@
+        @purchase_delivery_information.phone = '123-4567-8901'
+        @purchase_delivery_information.valid?
+        expect(@purchase_delivery_information.errors.full_messages).to include("Phone Input only number")
+      end
```
電話番号が11桁を超える場合もテストしましょう！

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

また、記述されたコードを拝見したところ、Formオブジェクトにpriceを渡す記述がされており、それに伴ってpriceのテストコードも記述されておりました。
アプリ自体は動くので問題ありませんが、priceは出品の段階で保存されているため、Formオブジェクトに渡す必要はありません。
もし不要な記述を減らしたいということであれば、priceについての記述は削除することをお勧めいたします。
こちらはあくまで任意ですので、必ずしも修正いただく必要はございません。

修正後、再度レビュー依頼フォームのご提出をよろしくお願いいたします。

tomo1107 commented:
> レビュー依頼ありがとうございます！
> コメントしましたので、ご確認をお願いいたします！
> 
> また、記述されたコードを拝見したところ、Formオブジェクトにpriceを渡す記述がされており、それに伴ってpriceのテストコードも記述されておりました。
> アプリ自体は動くので問題ありませんが、priceは出品の段階で保存されているため、Formオブジェクトに渡す必要はありません。
> もし不要な記述を減らしたいということであれば、priceについての記述は削除することをお勧めいたします。
> こちらはあくまで任意ですので、必ずしも修正いただく必要はございません。
> 
> 修正後、再度レビュー依頼フォームのご提出をよろしくお願いいたします。

ご確認頂きありがとうございます！

上記のpriceの件を含めて、全てのご指摘に対して是正対応を行いましたので、
再レビューの程、よろしくお願いします！

- テスト結果の画像
https://gyazo.com/5fe1fb7ee41e0e0af222e976a15314cb

chobi555 commented:
レビュー依頼ありがとうございます！
いただいたレビュー依頼ですが、修正点の差分があがっておりません。
お手数ですが、コミットとプッシュのし忘れを確認して修正点の差分を出して再度ご提出をお願いいたします。

tomo1107 committed:
商品購入機能-是正1

以下のご指摘に対して、是正措置を行いました。
https://github.com/tomo1107/furima-35073/pull/8#pullrequestreview-639589943

tomo1107 commented:
大変失礼いたしました。
変更点をコミットしましたので、レビューの程、よろしくお願いします！
https://github.com/tomo1107/furima-35073/pull/8#issuecomment-823058113

TakashiKozuma commented:
レビュー依頼ありがとうございます。
修正も含め全て確認が完了しました。
LGTMです！購入機能の実装お疲れさまでした。

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切ってから次にお進み下さい。
